multigeiger/tube.cpp

```diff
@@ -35,22 +36,133 @@ TUBETYPE tubes[] = {
 
 volatile bool isr_GMC_cap_full;
 
+volatile unsigned long isr_hv_pulses;
+volatile bool isr_hv_charge_error;
+
 volatile unsigned int isr_GMC_counts;
 volatile unsigned long isr_count_timestamp;
 volatile unsigned long isr_count_time_between;
 
-static unsigned long hvpulse_timestamp;  // initialized via setup_tube -> gen_charge_pulses
-
 // MUX (mutexes used for mutual exclusive access to isr variables)
 portMUX_TYPE mux_cap_full = portMUX_INITIALIZER_UNLOCKED;
 portMUX_TYPE mux_GMC_count = portMUX_INITIALIZER_UNLOCKED;
+portMUX_TYPE mux_hv = portMUX_INITIALIZER_UNLOCKED;
+
+// Maximum amount of HV capacitor charge pulses to generate in one charge cycle.
+#define MAX_CHARGE_PULSES 3333
+
+// hw timer period and microseconds -> periods conversion
+#define PERIOD_DURATION_US 100
+#define PERIODS(us) ((us) / PERIOD_DURATION_US)
+
+void IRAM_ATTR isr_recharge() {
+  // this code is periodically called by a timer hw interrupt, always same period.
+  // we need to decide internally whether we actually want to do something.
+  //
+  // note: this is implemented like it is because dynamically reprogramming the hw timer
+  // to a different period would require us to call library functions like timerAlarmWrite
+  // which are **not** in IRAM (but in flash) and doing that can lead to spurious fatal
+  // exceptions like "Cache disabled but cached memory region accessed".
```

Comment:
looks like this is the way to have reliable timer ISR code on ESP32.

Reply:
100us granularity so we could adjust these 1500us high 1000us low timings, if needed.

it's important to neither go way too fast here (too much interrupt overhead), nor too slow (to coarse granularity).